components/eamxx/src/physics/nudging/atmosphere_nudging.cpp

```diff
@@ -0,0 +1,264 @@
+      auto qv_bef_1d = ekat::subview(qv_bef,icol);
+      auto qv_aft_1d = ekat::subview(qv_aft,icol);
+      const auto range = Kokkos::TeamThreadRange(team, num_vert_packs);
+      Kokkos::parallel_for(range, [&] (const Int & k) {
+	  T_mid_1d(k)=w_bef*T_mid_bef_1d(k) + w_aft*T_mid_aft_1d(k);
+	  p_mid_1d(k)=w_bef*p_mid_bef_1d(k) + w_aft*p_mid_aft_1d(k);
+	  hw_2d(0,k)=w_bef*hw_bef_2d(0,k) + w_aft*hw_aft_2d(0,k);
+	  hw_2d(1,k)=w_bef*hw_bef_2d(1,k) + w_aft*hw_aft_2d(1,k);
+	  qv_1d(k)=w_bef*qv_bef_1d(k) + w_aft*qv_aft_1d(k);
+      });
+      team.team_barrier();
+  });
+  Kokkos::fence();   
+
+}
+  
+// =========================================================================================
+void NUDGING::run_impl (const int dt)
+{
+  using namespace scream::vinterp;
+
+  //Have to add dt because first time iteration is at 0 seconds where you will not have
+  //any data from the field. The timestamp is only iterated at the end of the
+  //full step in scream.
+  auto ts = timestamp()+dt;
+  auto time_since_zero = ts.seconds_from(ts0);
+  //Don't nudge for times less than first time step in nudged file to avoid
+  //extrapolation
+  if (time_since_zero < time_step_file){ return;}
+  //perform time interpolation
+  time_interpolation(time_since_zero);
+
+  auto T_mid       = get_field_out("T_mid").get_view<mPack**>();
+  auto qv          = get_field_out("qv").get_view<mPack**>();
+  auto hw          = get_field_out("horiz_winds").get_view<mPack***>();
+
+  const auto& p_mid    = get_field_in("p_mid").get_view<const mPack**>();
+  auto p_mid_v = view_Nd<mPack,2>("",p_mid.extent_int(0),p_mid.extent_int(1));
+  Kokkos::deep_copy(p_mid_v,p_mid);
+  
+  view_Nd<mPack,2> T_mid_out("T_mid_out",m_num_cols,T_mid.extent_int(1));
+  view_Nd<mPack,2> T_mid_ext_p("",T_mid_ext.extent_int(0),T_mid_ext.extent_int(1));
+  Kokkos::deep_copy(T_mid_ext_p,T_mid_ext);
+
+  view_Nd<mPack,2> qv_out("qv_out",m_num_cols,qv.extent_int(1));
+  view_Nd<mPack,2> qv_ext_p("qv_ext_p",qv_ext.extent_int(0),qv_ext.extent_int(1));
+  Kokkos::deep_copy(qv_ext_p,qv_ext);
+
+  view_Nd<mPack,3> hw_out("hw_out",m_num_cols,hw.extent_int(1),hw.extent(2));
+  view_Nd<mPack,3> hw_ext_p("hw_ext_p",
+  			    horiz_winds_ext.extent_int(0),
+  			    horiz_winds_ext.extent_int(1),
+  			    horiz_winds_ext.extent_int(2));
+  Kokkos::deep_copy(hw_ext_p,horiz_winds_ext);
+
+  view_Nd<mPack,2> p_mid_ext_p("",p_mid_ext.extent_int(0),p_mid_ext.extent_int(1));
+  Kokkos::deep_copy(p_mid_ext_p,p_mid_ext);
+
+  //Now perform the vertical interpolation from the external file to the internal
+  //field levels
+  perform_vertical_interpolation<Real,1,2>(p_mid_ext_p,
+                                           p_mid_v,
+                                           T_mid_ext_p,
+                                           T_mid_out,
+                                           m_num_src_levs,
+                                           m_num_levs);
+  Kokkos::deep_copy(T_mid,T_mid_out);
```

Comment:
I thought nudging was meant to give a little push toward the "reference" solution. Here, I see that we set the solution to _match_ the reference one. I guess I'm showing my newbiness here, but is that what nudging really is?

Reply:
Yes, it is meant to match the reference one not just give it a little push towards the reference one. Maybe it is a bit of a misnomer, but it is the nomenclature used in the literature.  You are only matching a few of the reference fields and keeping the rest of the internal fields unchanged, so maybe this is the explanation as to why it is denoted as nudging. You are replacing some part of the model but not the whole thing. Maybe there is a better name

Reply:
Thanks!